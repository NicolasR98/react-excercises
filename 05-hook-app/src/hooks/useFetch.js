import { useEffect, useState } from 'react';

export const useFetch = (url) => {
	const [state, setState] = useState({
		data: null,
		error: null,
		loading: true,
	});

	useEffect(() => {
		setState({
			loading: true,
			error: null,
			data: null,
		});

		fetch(url)
			.then((resp) => resp.json())
			.then((data) => {
				setState({
					loading: false,
					error: null,
					data,
				});
			});
	}, [url]);

	return state;
};
