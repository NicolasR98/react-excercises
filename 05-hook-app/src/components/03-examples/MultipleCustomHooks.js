import React from 'react';
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch';

export const MultipleCustomHooks = () => {
	const { counter, increment } = useCounter();
	const { loading, data } = useFetch(
		`https://breakingbadapi.com/api/quotes/${counter}`
	);

	const isValidData = !loading && data.length > 0;
	const { author, quote } = isValidData && data[0];

	return (
		<div>
			<h1>Breaking Bad quoutes</h1>
			<hr />

			{loading && (
				<div className="alert alert-info text-center">... Loading</div>
			)}

			{!loading && (
				<blockquote className="blockquote text-left">
					<p>{quote}</p>
					<footer className="blockquote-footer">{author}</footer>
				</blockquote>
			)}
			<button className="btn btn-primary" onClick={() => increment(1)}>
				Next quoute
			</button>
		</div>
	);
};
