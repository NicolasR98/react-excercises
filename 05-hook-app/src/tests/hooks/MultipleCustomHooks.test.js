import { render } from '@testing-library/react';
import { shallow } from 'enzyme';
import React from 'react';
import { MultipleCustomHooks } from '../../components/03-examples/MultipleCustomHooks';
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch';

jest.mock('../../hooks/useFetch');
jest.mock('../../hooks/useCounter');

describe('Tests on <MultipleCustomsHooks />', () => {
	beforeEach(() => {
		useCounter.mockReturnValue({
			counter: 1,
			increment: () => {},
		});
	});

	test('should display correctly', () => {
		useFetch.mockReturnValue({
			data: null,
			loading: true,
			error: null,
		});

		const { container } = render(<MultipleCustomHooks />);
		expect(container).toMatchSnapshot();
	});

	test('should display the info', () => {
		useFetch.mockReturnValue({
			data: [
				{
					author: 'Nicolas',
					quote: 'Hello world!',
				},
			],
		});

		const wrapper = shallow(<MultipleCustomHooks />);
		expect(wrapper.find('.alert').exists()).toBeFalsy();
		expect(wrapper.find('p').text().trim()).toBe('Hello world!');
		expect(wrapper.find('footer').text().trim()).toBe('Nicolas');
	});
});
