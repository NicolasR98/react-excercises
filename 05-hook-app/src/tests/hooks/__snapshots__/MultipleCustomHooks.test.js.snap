// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <MultipleCustomsHooks /> should display correctly 1`] = `
<div>
  <div>
    <h1>
      Breaking Bad quoutes
    </h1>
    <hr />
    <div
      class="alert alert-info text-center"
    >
      ... Loading
    </div>
    <button
      class="btn btn-primary"
    >
      Next quoute
    </button>
  </div>
</div>
`;
